Only store the MVCCMetadata value when there is an intent.

petermattis opened the pull request:
When an intent is not present the MVCCMetadata is implicit and is
reconstructed from the most recent versioned value. This provides a very
nice performance improvement:

```
name                                old time/op    new time/op    delta
MVCCScan1Version1Row8Bytes-8          6.04µs ± 1%    4.52µs ± 2%  -25.10%  (p=0.000 n=10+10)
MVCCScan1Version10Rows8Bytes-8        34.3µs ± 0%    21.9µs ± 0%  -35.99%    (p=0.000 n=9+9)
MVCCScan1Version100Rows8Bytes-8        302µs ± 0%     183µs ± 1%  -39.17%   (p=0.000 n=9+10)
MVCCScan1Version1000Rows8Bytes-8      2.95ms ± 0%    1.78ms ± 0%  -39.48%   (p=0.000 n=10+9)
MVCCScan10Versions1Row8Bytes-8        35.4µs ± 2%    29.9µs ± 0%  -15.49%   (p=0.000 n=10+9)
MVCCScan10Versions10Rows8Bytes-8       133µs ± 1%     108µs ± 0%  -18.56%   (p=0.000 n=10+8)
MVCCScan10Versions100Rows8Bytes-8     1.05ms ± 1%    0.84ms ± 1%  -19.71%  (p=0.000 n=10+10)
MVCCScan10Versions1000Rows8Bytes-8    10.1ms ± 0%     8.1ms ± 1%  -19.76%    (p=0.000 n=9+9)
MVCCGet1Version8Bytes-8               26.5µs ± 1%    20.6µs ± 1%  -22.18%   (p=0.000 n=9+10)
MVCCGet10Versions8Bytes-8             33.0µs ± 1%    30.2µs ± 1%   -8.26%   (p=0.000 n=10+9)
```

Adjusted the scan benchmarks to operate serially instead of in
parallel. Adjusted the data generation to be deterministic.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3324)

<!-- Reviewable:end -->

petermattis committed:
Only store the MVCCMetadata value when there is an intent.

When an intent is not present the MVCCMetadata is implicit and is
reconstructed from the most recent versioned value. This provides a very
nice performance improvement:

name                                old time/op    new time/op    delta
MVCCScan1Version1Row8Bytes-8          6.04µs ± 1%    4.52µs ± 2%  -25.10%  (p=0.000 n=10+10)
MVCCScan1Version10Rows8Bytes-8        34.3µs ± 0%    21.9µs ± 0%  -35.99%    (p=0.000 n=9+9)
MVCCScan1Version100Rows8Bytes-8        302µs ± 0%     183µs ± 1%  -39.17%   (p=0.000 n=9+10)
MVCCScan1Version1000Rows8Bytes-8      2.95ms ± 0%    1.78ms ± 0%  -39.48%   (p=0.000 n=10+9)
MVCCScan10Versions1Row8Bytes-8        35.4µs ± 2%    29.9µs ± 0%  -15.49%   (p=0.000 n=10+9)
MVCCScan10Versions10Rows8Bytes-8       133µs ± 1%     108µs ± 0%  -18.56%   (p=0.000 n=10+8)
MVCCScan10Versions100Rows8Bytes-8     1.05ms ± 1%    0.84ms ± 1%  -19.71%  (p=0.000 n=10+10)
MVCCScan10Versions1000Rows8Bytes-8    10.1ms ± 0%     8.1ms ± 1%  -19.76%    (p=0.000 n=9+9)
MVCCGet1Version8Bytes-8               26.5µs ± 1%    20.6µs ± 1%  -22.18%   (p=0.000 n=9+10)
MVCCGet10Versions8Bytes-8             33.0µs ± 1%    30.2µs ± 1%   -8.26%   (p=0.000 n=10+9)

Adjusted the scan benchmarks to operate serially instead of in
parallel. Adjusted the data generation to be deterministic.

petermattis commented:
Still working on 2 test failures (thus the `[ci skip]`), but I think the implementation is solid and it is just the tests that need fixing.

petermattis commented:
The tests are all passing now.

tschottdorf commented:
we also keep the meta key for deletion tombstones, right? If so, mind updating the commit message?

tschottdorf commented:
ah, I see that you reconstruct this from the length of the value. Does that affect how we see keys with a zero-length value?

petermattis commented:
@tschottdorf No, the `MVCCMetadata` for deletion tombstones is implicit as well. Deletion tombstones are stored as `<key>/<timestamp> -> nil`. It's straightforward to reconstruct the `MVCCMetadata` from this info.

petermattis commented:
Our messages crossed. A non-deleted key with a zero-length (user supplied) value has an actual value of `roachpb.Value{RawBytes: nil}`. A deleted key has an actual value of `nil`. Note that this was true even before this change and made possible the removal of `MVCCValue`.

tschottdorf commented:
LGTM, but I didn't hand-verify all the stats fallout. I hope our testing coverage is good there.

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/mvcc.go, line 852 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hZ-AYVtsVXKVCB4Ds:-K4hZ-AYVtsVXKVCB4Dt:816992505) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L852)):*
shouldn't you be able to remove the meta key in this case?

---

*[storage/engine/mvcc.go, line 1302 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hddYFDnNq7ZoyZxtL:-K4hddYFDnNq7ZoyZxtM:454781044) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L1302)):*
I think this TODO can go.

---

*[storage/engine/mvcc.go, line 1377 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4he8bh1sQ56EY-da6f:-K4he8bh1sQ56EY-da6g:-622175551) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L1377)):*
you already `Clear` higher up, but seems like doing it possibly twice is ok.

---

*[storage/engine/mvcc.go, line 1487 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4heaxH87wivvoh2Gqx:-K4heaxH87wivvoh2Gqy:-2032884603) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L1487)):*
comment is stale.

---

*[storage/engine/mvcc.go, line 1563 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4heft3_25Tvp1hUjBv:-K4heft3_25Tvp1hUjBw:2046421771) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L1563)):*
s/prevKey/lastKey/? I always see `prev` as the opposite of `next`.

---

*[storage/engine/rocksdb_test.go, line 205 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hfUyXOA5hnK-Yxvqg:-K4hfUyXOA5hnK-Yxvqh:1216400712) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/rocksdb_test.go#L205)):*
Is it guaranteed that writing more data will give the same number of SSTables? I have no good understanding of when RocksDB would create a new one.

---

*[storage/gc_queue.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hft9xkx20Qz-RaIIg:-K4hft9xkx20Qz-RaIIh:-140358876) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/gc_queue.go#L178)):*
does this work in case this is hit with an implicit meta key from l. 221 below? Is it exercised and does the right thing below? Sounds like you should be synthesizing a real meta key.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-:-K4hhgixkQl2iNXxdeG7:686224746)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I'm decently comfortable with the stats changes. I went through all of them in `storage/engine` and adjusted by hand in the tests and verified they matched my expectations.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/mvcc.go, line 852 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hZ-AYVtsVXKVCB4Ds:-K4hjQfAvo0HKur10afE:883324306) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L852)):*
Look closer. We're never adding the meta key in the first place. What this is doing is maintaining the invariant that even when the meta key is not present we account for the length of the key. The way stats were previously collected for versioned keys is that we account for the full key size on the meta key and only `mvccVersionTimestampSize` for the keys with timestamps. I've maintained that behavior by accounting for the full key size on the latest version (i.e. the one we implicitly reconstruct the metadata from). A bit subtle. I'll add a comment.

---

*[storage/engine/mvcc.go, line 1302 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hddYFDnNq7ZoyZxtL:-K4hk-HUOK44weSEgKzl:-1413315643) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L1302)):*
Yeah, I'm not even sure how such a merge operator would be written since the key would be changing.

---

*[storage/engine/mvcc.go, line 1377 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4he8bh1sQ56EY-da6f:-K4hkMMoRvl0B1_4pvXZ:21717270) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L1377)):*
I think we're only clearing once. Yes, there is another clear higher up in this method, but it is on a different code path.

---

*[storage/engine/mvcc.go, line 1487 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4heaxH87wivvoh2Gqx:-K4hkZTKI9g8yINbtndc:1758359711) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L1487)):*
Fixed.

---

*[storage/engine/mvcc.go, line 1563 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4heft3_25Tvp1hUjBv:-K4hkbzwytHD02vwpHbh:1660675680) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/mvcc.go#L1563)):*
Sure. Done.

---

*[storage/engine/rocksdb_test.go, line 205 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hfUyXOA5hnK-Yxvqg:-K4hkirmo6MjWI6wlLW2:597562673) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/rocksdb_test.go#L205)):*
RocksDB buffers writes into a memory table while also adding them to a WAL. When the memory table reaches a certain size it is flushed to disk creating a new sstable. So it isn't quite guaranteed that we'll end up with 20 sstables because the auto-flushing might have kicked in. I could actually turn that off. Not sure if it matters.

---

*[storage/gc_queue.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hft9xkx20Qz-RaIIg:-K4hl7811T6jtpSv9Awa:-455565103) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/gc_queue.go#L178)):*
Yes, it seems to work. The unmarshaling succeeds and creates an empty `MVCCMetadata`. I'll add a comment at line 221 explaining why it is safe.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-:-K4hlOuAyESjWUalNr8S:-1744272052)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 7 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/engine/rocksdb_test.go, line 205 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hfUyXOA5hnK-Yxvqg:-K4i7tUl1d7UrwkTmQYu:1300140805) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/rocksdb_test.go#L205)):*
What guarantees that we even get multiple ones? 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 7 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/engine/rocksdb_test.go, line 205 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324#-K4hfUyXOA5hnK-Yxvqg:-K4iH0yM8b1TUh5OFNuG:-1172629923) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f03ee62d1b8e87394114d9019f0879f9f063257/storage/engine/rocksdb_test.go#L205)):*
The call to `rocksdb.Flush` a few lines down explicitly creates an sstable from the current memtable. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3324)*
<!-- Sent from Reviewable.io -->
